backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/scheduling/utils/VdsCpuUnitPinningHelper.java

```diff
@@ -239,10 +194,39 @@ public int getDedicatedCount(Guid vdsId) {
     }
 
     private List<Integer> getOnlineSockets(List<VdsCpuUnit> cpuTopology) {
-        return cpuTopology.stream().map(VdsCpuUnit::getSocket).collect(Collectors.toList());
+        return cpuTopology.stream().map(VdsCpuUnit::getSocket).distinct().collect(Collectors.toList());
     }
 
     private List<Integer> getOnlineCores(List<VdsCpuUnit> cpuTopology, int socket) {
-        return getCoresInSocket(cpuTopology, socket).stream().map(VdsCpuUnit::getCore).collect(Collectors.toList());
+        return getCoresInSocket(cpuTopology, socket).stream().map(VdsCpuUnit::getCore).distinct().collect(Collectors.toList());
+    }
+
+    private List<Integer> getOnlineCores(List<VdsCpuUnit> cpuTopology) {
+        return cpuTopology.stream().map(VdsCpuUnit::getCore).distinct().collect(Collectors.toList());
+    }
+
+    private List<VdsCpuUnit> getMaxFreedSocket(List<VdsCpuUnit> cpuTopology) {
+        List<VdsCpuUnit> chosenSocket = Collections.emptyList();
+        List<VdsCpuUnit> temp;
+        for (int socket : getOnlineSockets(cpuTopology)) {
+            temp = getFreeCpusInSocket(cpuTopology, socket);
+            if (temp.size() > chosenSocket.size()) {
+                chosenSocket = temp;
+            }
+        }
+        return chosenSocket;
+    }
+
+    private int getAvailableCores(List<VdsCpuUnit> cpuTopology, int socket, int vThreads) {
+        int count = 0;
+        for (int core : getOnlineCores(cpuTopology, socket)) {
+            List<VdsCpuUnit> freeCpusInCore = getFreeCpusInCore(cpuTopology, socket, core);
+            int coreThreads = freeCpusInCore.size();
+            while (coreThreads >= vThreads) {
+                count++;
+                coreThreads -= vThreads;
+            }
```

Comment:
that's the last part where we still have you replace those 4-lines with a single one?